pkg/contract/contract.go

```diff
@@ -1,225 +1,116 @@
-			Version:            "v2",
-			Language:           "EN",
-			SignerAttributes:   StandardSignerAttributes,
-			Template:           `EN:PractitionerLogin:v2 Undersigned gives permission to {{acting_party}} to make request to the Nuts network on behalf of {{legal_entity}} and itself. This permission is valid from {{valid_from}} until {{valid_to}}.`,
-			TemplateAttributes: []string{"acting_party", "legal_entity", "valid_from", "valid_to"},
-			Regexp:             `EN:PractitionerLogin:v2 Undersigned gives permission to (.+) to make request to the Nuts network on behalf of (.+) and itself. This permission is valid from (.+) until (.+).`,
-		},
-	}},
-}
-
-// NewFromMessageContents finds the contract for a certain message.
-// Every message should begin with a special sequence like "NL:ContractName:version".
-func NewFromMessageContents(contents string, validContracts Matrix) (*Template, error) {
-	r, _ := regexp.Compile(`^(.{2}):(.+):(v\d+)`)
-
-	matchResult := r.FindSubmatch([]byte(contents))
-	if len(matchResult) != 4 {
-		return nil, fmt.Errorf("%w: could not extract contract version, language and type", ErrInvalidContractText)
+	// first, find the contract Template
+	template, err := validContracts.FindFromRawContractText(rawContractText)
+	if err != nil {
+		return nil, err
 	}
 
-	language := Language(matchResult[1])
-	contractType := Type(matchResult[2])
-	version := Version(matchResult[3])
-
-	return NewByType(contractType, language, version, validContracts)
-
-}
-
-// NewByType returns the contract for a certain type, language and version. If version is omitted "v1" is used
-// If no contract is found, the error vaule of ErrContractNotFound is returned.
-func NewByType(contractType Type, language Language, version Version, validContracts Matrix) (*Template, error) {
-	if version == "" {
-		version = "v1"
-	}
-	if contract, ok := validContracts[language][contractType][version]; ok {
-		return contract, nil
+	contract := &Contract{
+		Template:        template,
+		RawContractText: rawContractText,
 	}
 
-	return nil, fmt.Errorf("type %s, lang: %s, version: %s: %w", contractType, language, version, ErrContractNotFound)
-}
-
-func (c Template) timeLocation() *time.Location {
-	loc, _ := time.LoadLocation("Europe/Amsterdam")
-	return loc
-}
-
-func (c Template) RenderTemplate(vars map[string]string, validFromOffset, validToOffset time.Duration) (string, error) {
-	vars["valid_from"] = monday.Format(time.Now().Add(validFromOffset).In(c.timeLocation()), timeLayout, monday.LocaleNlNL)
-	vars["valid_to"] = monday.Format(time.Now().Add(validToOffset).In(c.timeLocation()), timeLayout, monday.LocaleNlNL)
+	if err = contract.extractParams(); err != nil {
+		return nil, err
+	}
 
-	return mustache.Render(c.Template, vars)
+	return contract, nil
 }
 
-func (c Template) ExtractParams(text string) (map[string]string, error) {
-	r, _ := regexp.Compile(c.Regexp)
-	matchResult := r.FindSubmatch([]byte(text))
+func (sc *Contract) extractParams() error {
```

Comment:
I don't like mutating functions that return nothing (other than an error) especially when it's not 100% clear from the function name. `extractParam` makes me think it just extracts and returns stuff. Maybe change it so that it just returns the parameters (non-mutating) and assign them in the calling function?

Otherwise make it more clear that it should be called to make the Contract state valid (when/why?).

Reply:
I see your point, but its an internal function, just to extract the param extraction from the initializer method. I renamed it to `initParams`.